modin/utils.py

```diff
@@ -58,15 +64,143 @@ def _get_indent(doc: str) -> int:
     int
         Minimal indent (excluding empty lines).
     """
+    indents = _get_indents(doc)
+    return min(indents) if indents else 0
+
+
+def _get_indents(source: Union[list, str]) -> list:
+    """
+    Compute indentation for each line of the source string.
+
+    Parameters
+    ----------
+    source : str or list of str
+        String to compute indents for. Passed list considered
+        as a list of lines of the source string.
+
+    Returns
+    -------
+    list of ints
+        List containing computed indents for each line.
+    """
     indents = []
-    for line in doc.splitlines():
+
+    if not isinstance(source, list):
+        source = source.splitlines()
+
+    for line in source:
         if not line.strip():
             continue
         for pos, ch in enumerate(line):
             if ch != " ":
                 break
         indents.append(pos)
-    return min(indents) if indents else 0
+    return indents
+
+
+def format_string(template: str, **kwargs) -> str:
+    """
+    Insert passed values at the corresponding placeholders of the specified template.
+
+    In contrast with the regular ``str.format()`` this function computes proper
+    indents for the placeholder values.
+
+    Parameters
+    ----------
+    template : str
+        Template to substitute values in.
+    **kwargs : dict
+        Dictionary that maps placeholder names with values.
+
+    Returns
+    -------
+    str
+        Formated string.
+    """
+    # We want to change indentation only for those values, whose placeholder located
+    # at the beginning of the line, in that case, the placeholder sets an indentation
+    # that the filling value has to obey.
+    template_lines = template.splitlines()
+    # RegExp determining placeholders located at the beginning of the line.
+    regex = r"^[ ^{]*{(\w+)}"
+    to_align = OrderedDict()
+    for line_idx, line in enumerate(template_lines):
+        if line.strip() == "":
+            continue
+        match = re.search(regex, line)
+        if match is None:
+            continue
+        value_key = match.group(1)
+        to_align[value_key] = line_idx
+
+    indents = _get_indents([template_lines[i] for i in to_align.values()])
+    for key, nspaces in zip(to_align.keys(), indents):
+        value = kwargs.get(key)
+        if not value:
+            continue
+        value = dedent(value)
+
+        # Since placeholder is located at the beginning of a new line, its already
+        # has '\n' before it, so to avoid double new lines we want to ensure that
+        # the value string doesn't starts with '\n'
+        if value[0] == "\n":
+            value = value[1:]
```

Comment:
```suggestion
        value = value.lstrip("\n")
```
?

Reply:
My explanation at the above comment wasn't clear enough, we don't want to omit all of the leading '\n' characters, otherwise, it'll be impossible to substitute a value with the leading empty line.

<details><summary>Example</summary>

```python
from modin.utils import format_string

value1 = """
    I don't want to have leading '\\n' when this value
    will be substituted
"""

value2 = """

    And here I want to have leading '\\n' because I left
    an explicit empty line at the beginning
"""

template = """
    Placeholder1:
    {placeholder1}
    Placeholder2:
    {placeholder2}
"""

print(format_string(template, placeholder1=value1, placeholder2=value2))

# Output:
#   Placeholder1:
#   I don't want to have leading '\n' when this value
#   will be substituted
#
#   Placeholder2:
#
#   And here I want to have leading '\n' because I left
#   an explicit empty line at the beginning
#

```

</details>